lib/types.js

```diff
@@ -1226,6 +1243,44 @@ UnionType.prototype._branchConstructor = function () {
   throw new Error('unions cannot be directly wrapped');
 };
 
+
+function generateProjectionIndexer(projectionFn) {
+  return (val) => {
+    const index = projectionFn(val);
+    if (typeof index !== 'number') {
+      throw new Error(`Projected index '${index}' is not valid`);
+    }
+    return index;
+  };
+}
+
+function generateDefaultIndexer() {
```

Comment:
this might be better as a private function on the class, given it still uses class context for `this._bucketIndices` and `this._getBranchIndex`.

Reply:
The idea here is to move default indexing logic out of `UnwrappedUnion` into this indexer. We can move the method out and replace the fields with local variables. Something like:

```javascript
function generateDefaultIndexer(types) {
  let dynamicBranches = null;
  let bucketIndices = {};

  function getBranchIndex (any, index) {
    // Equivalent logic to _getBranchIndex...
  }

  types.forEach(function (type, index) {
    // Same as currently, replacing this._dynamicBranches with
    // dynamicBranches and similar for bucketIndices. 
  });

  return (val) => {
    // Same as in the forEach above here.
  };
}
```

We also need to perform a couple small changes in `_copy` to replace the accesses to `_bucketIndices`. For example:

```javascript
case 1: // Here
  if (isJsonBuffer(val)) {
    let bufIndex = this.types.findIndex(t => getTypeBucket(t) === 'buffer');
    if (bufIndex >= 0) {
      index = bufIndex;
    }
  }
  index ??= this._getIndex(val);
  break;
case 2:
  if (val === null) {
    index = this._getIndex(null); // And here
    // ...
```

We could optimize `case 1` to avoid `findIndex` calls but that can wait. We may remove the `coerceBuffers` option altogether as discussed [here](https://github.com/mtth/avsc/pull/452#issuecomment-2342308104).

Reply:
bd2c6dd